### ISSUE TITLE
Building on debian with GCC 6 (and/or updated binutils, maybe?) fails during check-security-hardening.sh

### ISSUE THREAD

**Issue by bitcartel** [zcash/zcash#2212]
http://ci.z.cash:8010/builders/zcash/builds/391/steps/full-test-suite.sh/logs/stdio

```
===== BEGIN: /home/admin/bbs/zcash/build/qa/zcash/check-security-hardening.sh
make: Entering directory '/home/admin/bbs/zcash/build/src'
Checking binary security of [zcashd zcash-cli zcash-tx]...
zcash-tx: failed RELRO
Makefile:5791: recipe for target 'check-security' failed
make: *** [check-security] Error 1
make: Leaving directory '/home/admin/bbs/zcash/build/src'
===== FAILED: /home/admin/bbs/zcash/build/qa/zcash/check-security-hardening.sh
```

**Comment by bitcartel** [comment]
Hi @defuse :-)

**Comment by daira** [comment]
[Conversation on #zcash-dev](https://chat.zcashcommunity.com/channel/zcash-dev?msg=vGXo6y2PhxwYZz4AT):
> kevin: [...] I'm building in a fresh Jessie vm
[That build worked.]
> daira: ok
> daira: does the VM do an 'apt-get update'?
> kevin: yeah, everything is latest from stable repos
> daira: ok, so those repos could have changed
> kevin: same with AMIs we used to build, fully updated
> daira: it could be something that was updated when we tried gcc 6, but is *also* now updated in Jessie
> kevin: would any of these packages affect it?
```
root@ip-172-31-26-222:/etc/apt/preferences.d# apt-show-versions | grep -v jessie
autoconf:all 2.69-10 newer than version in archive
automake:all 1:1.15-6 newer than version in archive
binutils:amd64 2.28-2 newer than version in archive
bsdmainutils:amd64 9.0.12 newer than version in archive
ccache:amd64 3.3.4-1 newer than version in archive
geoip-database:all 20160812-1~bpo8+1 newer than version in archive
geoip-database-extra:all 20160812-1~bpo8+1 newer than version in archive
git:amd64 1:2.11.0-2 newer than version in archive
git-man:all 1:2.11.0-2 newer than version in archive
libgeoip1:amd64 1.6.7-2~bpo8+1 newer than version in archive
libncurses5:amd64 6.0+20161126-1 newer than version in archive
libncurses5-dev:amd64 6.0+20161126-1 newer than version in archive
libncursesw5:amd64 6.0+20161126-1 newer than version in archive
libp11-kit0:amd64 0.23.3-5 newer than version in archive
libpython-dev:amd64 2.7.13-2 newer than version in archive
libpython-stdlib:amd64 2.7.13-2 newer than version in archive
libpython2.7:amd64 2.7.13-2 newer than version in archive
libpython2.7-dev:amd64 2.7.13-2 newer than version in archive
libpython2.7-minimal:amd64 2.7.13-2 newer than version in archive
libpython2.7-stdlib:amd64 2.7.13-2 newer than version in archive
libreadline7:amd64 7.0-2 installed: No available version in archive
libssl1.1:amd64 1.1.0e-1 installed: No available version in archive
libtasn1-6:amd64 4.10-1 newer than version in archive
libtinfo-dev:amd64 6.0+20161126-1 newer than version in archive
libtinfo5:amd64 6.0+20161126-1 newer than version in archive
linux-libc-dev:amd64 4.9.16-1 newer than version in archive
m4:amd64 1.4.18-1 newer than version in archive
ncurses-bin:amd64 6.0+20161126-1 newer than version in archive
pkg-config:amd64 0.29-4+b1 newer than version in archive
python:amd64 2.7.13-2 newer than version in archive
python-cffi-backend:amd64 1.4.2-2~bpo8+1 installed: No available version in archive
python-cryptography:amd64 1.3.4-1~bpo8+2 newer than version in archive
python-dev:amd64 2.7.13-2 newer than version in archive
python-idna:all 2.0-3~bpo8+1 installed: No available version in archive
python-ipaddress:all 1.0.16-1~bpo8+1 installed: No available version in archive
python-minimal:amd64 2.7.13-2 newer than version in archive
python-ndg-httpsclient:all 0.4.2-1~bpo8+1 newer than version in archive
python-openssl:all 16.0.0-1~bpo8+1 newer than version in archive
python-pkg-resources:all 20.10.1-1.1~bpo8+1 newer than version in archive
python-pyasn1:all 0.1.9-1~bpo8+1 newer than version in archive
python-setuptools:all 20.10.1-1.1~bpo8+1 newer than version in archive
python-six:all 1.9.0-3~bpo8+1 newer than version in archive
python2.7:amd64 2.7.13-2 newer than version in archive
python2.7-dev:amd64 2.7.13-2 newer than version in archive
python2.7-minimal:amd64 2.7.13-2 newer than version in archive
unzip:amd64 6.0-21 newer than version in archive
zlib1g:amd64 1:1.2.8.dfsg-5 newer than version in archive
zlib1g-dev:amd64 1:1.2.8.dfsg-5 newer than version in archive
```
> daira: _reads_
> daira: binutils certainly could
> kevin okay.
> daira: that one jumps out at me, actually
> daira linux-libc-dev could, as well
> kevin: okay, I'll downgrade those
> _daira looks at how the relro check in ``security-hardening.sh`` works_
> bitcartel: I just scanned the scripts in that folder, no mention of gcc.
>
> daira:
```
# PIE, RELRO, Canary, and NX are tested by make check-security.
make -C "$REPOROOT/src" check-security
```
> kevin: okay, it passed on a fresh jessie VM
> kevin: it must be one of those newer packages
> daira: in ``src/Makefile.am``:
```
check-security: $(bin_PROGRAMS)
if HARDEN
        @echo "Checking binary security of [$(bin_PROGRAMS)]..."
        $(AM_V_at) READELF=$(READELF) OBJDUMP=$(OBJDUMP) $(top_srcdir)/contrib/devtools/$
endif

%.pb.cc %.pb.h: %.proto
        @test -f $(PROTOC)
        $(AM_V_GEN) $(PROTOC) --cpp_out=$(@D) --proto_path=$(abspath $(<D) $<)
```
> kevin: I'm cutting new AMIs that are straight jessie with no mixing right now
> daira: we'd better be careful that that isn't missing possible security fixes that could affect the build output
> daira: (but carry on cutting those AMIs)
> _daira reads the binutils 2.28-2 release notes_
> daira: https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=blob_plain;f=binutils/NEWS;hb=refs/tags/binutils-2_28
> daira: https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=blob_plain;f=gas/NEWS;hb=refs/tags/binutils-2_28
> daira: https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=blob_plain;f=ld/NEWS;hb=refs/tags/binutils-2_28
[snip discussion of printing out tool versions, which is #2218.]
> daira: there were changes to relro support in binutils 2.27: https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=blob_plain;f=ld/NEWS;hb=refs/tags/binutils-2_28
> daira: what's the binutils package version in stock Jessie with no updates?
> kevin: 2.25-5
> kevin: it's like the failure comes with a combination of packages -- newer binutils + older gcc? but with newer binutils and newer gcc, or older binutils and older gcc, it's fine
> daira: oh, that could make sense
> daira: if they made an incompatible change to binutils and a compensating change to gcc

**Comment by daira** [comment]
We need to fix this if we want to include AMQP in 1.0.9 (#2189).

**Comment by bitcartel** [comment]
@daira Resolved, see latest commit and comment in PR.  local testing passed, zkbot try passed, now merging.

**Comment by daira** [comment]
+1 for bumping this [until Stretch is released](https://lists.debian.org/debian-devel-announce/2015/03/msg00002.html).

**Comment by daira** [comment]
Building with GCC 6 or later is becoming increasingly common, so we should really fix this.

**Comment by str4d** [comment]
This was fixed by #2752 ([before](https://ci.z.cash/#/builders/12/builds/107), [after](https://ci.z.cash/#/builders/12/builds/110)).
